### ISSUE TITLE
Speedup 20-30% RNN

### ISSUE THREAD

**Issue by lpigou** [craffel/nntools#20]
In the CustomRecurrentLayer, if you calculate self.input_to_hidden.get_output(layer_input) outside the step function, I observed a GPU speedup of 20-30% for my RNN (bidirectional, 3 layers deep, 256 hidden states each layer, minibatch: 20, time length: 1000, features: 180).

Something like this:

```
    n_time_steps, n_batch = input.shape[0], input.shape[1]
    input = input.reshape((n_time_steps*n_batch, input.shape[2]))
    input = self.input_to_hidden.get_output(input)
    input = input.reshape((n_time_steps,n_batch,-1))

    # Create single recurrent computation step function
    def step(layer_input, hid_previous):
        return self.nonlinearity(
            layer_input +
            self.hidden_to_hidden.get_output(hid_previous))
```

**Comment by craffel** [comment]
Oh, good call.  We do this in the `LSTMLayer`, we should do this in the `CustomRecurrentLayer` too.

**Comment by JackKelly** [comment]
@lpigou this sounds like a really useful tweak.  I've tried to implement it and I'm getting some odd errors(*).  Please could you share the code for your entire `CustomRecurrentLayer.get_output_for()` method?  Or maybe, if @craffel is happy with this suggestion, please could you submit a pull request?

(*) A little more detail on the errors I was getting.  

Below is here's my version of `CustomRecurrentLayer.get_output_for()` ([this code is also in my fork of craffel/nntools](https://github.com/JackKelly/nntools/blob/speedup_issue20/lasagne/layers/recurrent.py#L105)):

``` python
def get_output_for(self, input, mask=None, *args, **kwargs):
    if input.ndim > 3:
        input = input.reshape((input.shape[0], input.shape[1],
                               T.prod(input.shape[2:])))

    # Input should be provided as (n_batch, n_time_steps, n_features)
    # but scan requires the iterable dimension to be first
    # So, we need to dimshuffle to (n_time_steps, n_batch, n_features)
    input = input.dimshuffle(1, 0, 2)

    # Computing input_dot_W outside of the `step` function gives a 
    # 20-30% speedup.
    n_time_steps, n_batch, n_features = input.shape
    input = input.reshape((n_time_steps * n_batch, n_features))
    input_dot_W = self.input_to_hidden.get_output(input)
    input_dot_W = input_dot_W.reshape((n_time_steps, n_batch, -1))

    # Create single recurrent computation step function
    def step(input_dot_W_n, hid_previous):
        return self.nonlinearity(
             input_dot_W_n +
             self.hidden_to_hidden.get_output(hid_previous))

    def step_masked(input_dot_W_n, mask, hid_previous):
        # If mask is 0, use previous state until mask = 1 is found.
        # This propagates the layer initial state when moving backwards
        # until the end of the sequence is found.
        hid = (step(input_dot_W_n, hid_previous) * mask
               + hid_previous * (1 - mask))
        return [hid]

    if self.backwards and mask is not None:
        mask = mask.dimshuffle(1, 0, 'x')
        sequences = [input_dot_W, mask]
        step_fun = step_masked
    else:
        sequences = input_dot_W
        step_fun = step

    output = theano.scan(step_fun, sequences=sequences,
                         go_backwards=self.backwards,
                         outputs_info=[self.hid_init],
                         truncate_gradient=self.gradient_steps)[0]

    # Now, dimshuffle back to (n_batch, n_time_steps, n_features))
    output = output.dimshuffle(1, 0, 2)

    if self.backwards:
        output = output[:, ::-1, :]

    return output
```

This works fine for `lasagne/examples/recurrent.py`.  But when I run it on my research model, I get an error.  My research model has the following parameters:

```
seq_length = 514
n_batch = 64 
num_units = 40
number of incomming features = 40
gradient_steps = 100
```

This is the error I get:

```
ValueError: dimension mismatch in args to gemm (40,6400)x(32896,40) -> (40,40)
Apply node that caused the error: GpuDot22(GpuFlatten{2}.0, GpuReshape{2}.0) 

Inputs types: [CudaNdarrayType(float32, matrix), CudaNdarrayType(float32, matrix)]
Inputs shapes: [(40, 6400), (32896, 40)]
# 40 = num_units
# 6400 = gradient_steps x n_batch = 100 x 64
# 32896 = seq_length x n_batch  = 514 x 64
Inputs strides: [(6400, 1), (40, 1)] 
Inputs values: ['not shown', 'not shown']
```

To give some context to those numbers:

The really weird part is that this code runs if I set `optimizer=None` in my `.theanorc`.

**Comment by JackKelly** [comment]
aha, my code runs if I don't set `gradient_steps` for the `RecurrentLayer`.

is there a way to make this code make when `gradient_steps` is set?  I can't think of a way, but I'm only just starting to understand Theano!

**Comment by craffel** [comment]
Hm, what I think is happening is that when the `step` function looks as it currently does

```
        def step(layer_input, hid_previous):
            return self.nonlinearity(
                self.input_to_hidden.get_output(layer_input) +
                self.hidden_to_hidden.get_output(hid_previous))
```

`Theano` just backpropagates this step function `gradient_steps` times, so the shapes are all correct, whereas with your

```
    input = input.reshape((n_time_steps * n_batch, n_features))
    input_dot_W = self.input_to_hidden.get_output(input)
    input_dot_W = input_dot_W.reshape((n_time_steps, n_batch, -1))
```

the number of steps computed in backpropagation will not be truncated.  The tricky thing is that I can't think of a way to both forward-propagate for all timesteps and truncate backpopagation with pre-computing the input mix.

**Comment by JackKelly** [comment]
> The tricky thing is that I can't think of a way to both forward-propagate
> for all timesteps and truncate backpopagation with pre-computing the input mix.

Glad it's not just me who can't figure out how to do it!  I think I'll revert my code back to the 'old' way of doing it (at least, for now).  Thanks for the quick reply.

**Comment by craffel** [comment]
Unless we can think of a way to do it (I will think about it when I loop back on Lasagne RNN stuff next month), the best solution I can think of would be to add a flag for `precompute_input` (we were thinking of doing this already) and then throw an error when both `gradient_steps` is set and `precompute_input=True`.

**Comment by JackKelly** [comment]
> best solution I can think of would be to add a flag for precompute_input (we were thinking of doing this already) and then throw an error when both gradient_steps is set and precompute_input=True.

I've had a shot at implementing that:

https://github.com/JackKelly/nntools/blob/speedup_issue20/lasagne/layers/recurrent.py

I haven't tested it thoroughly yet.

**Comment by craffel** [comment]
Yes, I think that will work.  We should also have this flag in `LSTMLayer`, I think - the original motivation was that precomputing the input in the `LSTMLayer` roughly quadruples the memory consumption, so it should be optional.

**Comment by craffel** [comment]
Fixed in #27 and 9def279.
